### DISCUSSION TITLE
Query on Iterative algorithm in Raman solver class

### DISCUSSION THREAD

**Discussion by mukherjeearup** [Telecominfraproject/oopt-gnpy#474] [Q&A] [answered]
Hi,
Many thanks on resolving my previous queries. Now I have a doubt on Iterative algorithm method in Raman solver class. 
In absence of Raman pumps, "First order derivative solution" is used for the evolution of power profile along the fiber which is clear. 

Now, when Raman pumps are present (Co/counter or both), both "First order derivative solution" as well as "Iterative algorithm" are used. "First order derivative solution" is used to get the initial co-propagating /counter propagating Raman pumps power profiles, and those initial co/counter profiles are further used in "iterative algorithm" method to get overall power profile (including signal wavelngths and Raman pumps wavelengths). Can you please explain more about this iterative algorithm and any ref paper followed for this method? 

Regards,
Arup

**Comment by mukherjeearup** [comment]
Please help @AndreaDAmico @jktjkt

**Comment by AndreaDAmico** [comment]
Dear @mukherjeearup, 
`RamanSolver.iterative_algorithm` is called only if both co- and counter-propagating powers are present (see gnpy.core.science_utils.py line 150):
`                # Co-propagating and Counter-propagating Profile Computation`
`                if co_frequency.size and cnt_frequency.size:`
It has to be notice that "co_frequency" includes the frequencies of both the propagating spectrum and the co-propagating Raman pumps.
Therefore, all the possible scenarios are the following:

1. Only propagating spectrum or Only Co-propagating Raman pumps or Only Counter-propagating Raman pumps--> Only `RamanSolver.first_order_derivative_solution`
3. Propagating spectrum and Co-propagating Raman pumps --> Only `RamanSolver.first_order_derivative_solution`
4. Propagating spectrum and Counter-propagating Raman pumps --> `RamanSolver.first_order_derivative_solution` and `RamanSolver.iterative_algorithm`
5. Propagating spectrum, Co- and Counter-propagating Raman pumps --> `RamanSolver.first_order_derivative_solution` and `RamanSolver.iterative_algorithm`
6. Co- and Counter-propagating Raman pumps --> `RamanSolver.first_order_derivative_solution` and `RamanSolver.iterative_algorithm`

The main point is that when both co- and counter-propagating powers are present, the stimulated Raman scattering effect is the solution of a differential equation with two points boundary conditions (namely the two fiber terminations). 
The `RamanSolver.iterative_algorithm` assumes a weak Raman interaction between Co- and Counter-propagating powers, where, the fiber span is long enough such that when Co-propagating powers have high values, Counter-propagating powers have low values, and _vice versa_. In such conditions, the SRS solution can be found as result of successive iterations integrating back and forth the first order differential equation, in turn, varying the co-propagating powers with fixed counter-propagating powers and _vice versa_.
We verified that this methodology converges to a solution of the SRS differential equations for realistic fiber span lengths and propagating power values.

**Reply by mukherjeearup** [reply]
Dear @AndreaDAmico ,
Thank you so much for your explanation. It helps :)

Regards,
Arup

**Comment by caffery-chen** [comment]
Dear Andrea,

I kinda of fell that the solving of the ODE for power evolution is based on Euler method, am I right?
Does this back and forth propagation guarantee a convergence?

Thanks.
Caffery

**Reply by AndreaDAmico** [reply]
Dear @caffery-chen,
You are right, in GNPy the SRS ODE for power evolution are solved with the Euler method.
As a matter of fact, the back and forth propagation does not guarantee the convergence (therefore we set a maximum number of iterations). On the other hand, we have verified that, under weak Raman interaction between Co- and Counter-propagating powers, the back and forth propagation method converge to a solution of the two boundary condition SRS ODE. We are exploring a formal solution of this problem based on a perturbative solution of the SRS ODE (https://arxiv.org/abs/2304.11756).

**Reply by caffery-chen** [reply]
Dear Andrea,
So far it seems the algo converges pretty well with the pump configruation I provided and only counter-prop is involved.
May I ask will the RK4 method provide faster convergence or maybe more accurate results?
Of course, reduce the step length of the Euler method might provide the same results.

Thanks.
Caffery

**Comment by mukherjeearup** [comment]
Dear @AndreaDAmico,
Many thanks for your previous answers. I have one doubt while calculating Raman amplifier NF by using "calculate_spontaneous_raman_scattering' function.

I am using only 5 counter propgating Raman pumps and using "calculate_spontaneous_raman_scattering' function to calculate ASE generated at signals (48 C band ch and 48 L band chs).

Then, I am using below formula to calculate effective NF at all channels. I am expecting effective NF to be negative, but always getting positive values from the below equation:

effective_NF = lin2db((Raman_ase/(h*baud_rate*on_off_gain*co_frequency_in_THz*1e12)+(1/on_off_gain)))

Here, Raman_ase is the output of "calculate_spontaneous_raman_scattering' function,
         baud rate  = 32 GBd
         on_off_gain =  calculating gain at the end of span by taking difference of power profile when pumps are On vs pumps are off
        co_frequency_in_THz =  signal frequencies in THz

could you please help how to arrive at effective NF from the Raman "calculate_spontaneous_raman_scattering' function?

Regards,
Arup

**Reply by mukherjeearup** [reply]
sorry the previous formula got edited after posting, so rewriting effective NF formula again:
effective_NF = lin2db((Raman_ase/(h x baud_rate x on_off_gain x co_frequency_in_THz x 1e12)+(1/on_off_gain)))

could you please help how to arrive at effective NF from the Raman "calculate_spontaneous_raman_scattering' function?

**Reply by mukherjeearup** [reply]
Dear @AndreaDAmico , just a gentle reminder if you can help me resolving my above query of Raman effective NF derivation from the output of "calculate_spontaneous_raman_scattering" function defined under Raman solver class

**Reply by AndreaDAmico** [reply, marked as answer]
Dear @mukherjeearup, 
Sorry for the late answer. Using your variable definition, the correct formula to evaluate the effective noise figure should be just the following:

effective_NF = lin2db{Raman_ase/[h x baud_rate x (on_off_gain -1) x co_frequency_in_THz x 1e12]}

Indeed, if the on_off_gain is significantly larger than one, you can remove the -1 term obtaining:

effective_NF = lin2db[Raman_ase/[h x baud_rate x on_off_gain x co_frequency_in_THz x 1e12]}

**Comment by caffery-chen** [comment]
Dear @AndreaDAmico,

We did measurements over 150km SMF28e with four counter-prop pumps, May I ask if the following logic is correct?

The on-off gain is calculated as following:
1. Let the C-Band signal propagate over the SSMF with Raman Flag 'on' and save the final si_ra_off.
2. Let the C-Band signal and Raman pump propgate over RamanFiber and save the final si_ra_on.
3. Calculate on-off gain:
             G_onoff = 10 * np.log10((si_ra_on.signal+si_ra_on.nli+si_ra_on.ase) / (si_ra_off.signal+si_ra_off.nli+si_ra_off.ase))
4. Compare with real measurements.

Not sure if the above logic is correct to get the on-off gain in the simulation?

One more thing is why the output from fiber.cr(frequency) seems to be a symmetric matrix wrt. diagonal?

![image](https://github.com/Telecominfraproject/oopt-gnpy/assets/12408183/2ec6eafe-d36d-4518-980c-2600b747a6b9)
From the equation (I hope I am citing the correct one), it seems that no-particular symmetry exists for the Cr calculation.

Thanks.
Caffery

**Reply by AndreaDAmico** [reply]
Dear @caffery-chen,

Thanks you for your update. The logic you described to get the on-off gain in the simulation is correct. Without the simulation configuration and the measurement data, it is hard to figure out the issue on my end. Let me try to list a few of possible subtle details in the Raman amplification scenarios that can affect the accuracy of the simulations:
- The input connection loss, the lumped loss along the fiber as well as the Raman pump connector loss are of course very important in this transmission scenario.
- GNPy assumes that the Raman pumps (as well as the signal) are completely unpolarized as we discussed in  [#477 ](https://github.com/Telecominfraproject/oopt-gnpy/discussions/477). This is not always true, and it can greatly affect the power transfer among pumps near the injection point.
- It may be not significant, but I want to emphasize that the GNPy implementation of the ASE generation due to the Raman amplification has room for improvement.

I hope that these comments are helpful. If you feel comfortable to share additional information and data, I will be happy to have a look into the problem and provide more help if possible. 

Regarding the symmetry of the Raman coefficient matrix, the vibrational loss (the v_i / v_j term that makes Eq. (6) asymmetric) is already considered in GNPy. The Raman coefficient is virtually more accurate than Eq. (6) because it includes the effective area scaling.  In general, these asymmetries are only significant over a wide frequency bandwidth.

Best,
Andrea

**Reply by caffery-chen** [reply]
Dear @AndreaDAmico,

APOLOGIZE for claiming a mismatch between the your Raman code and the measurement. 
I have removed the errorneous statement out of the previous comments.

The fault is on our side that we have a mis-communication in the actual pump laser wavelengths.
After I put the correct pump lasers's lambda, the results show very good match in terms on-off gain wrt. experiment.

Even with gnpy 2.7.0, the results look very good.
I will further evaluate with GNPY 2.9.0.

Thanks for your feedback again.

Caffery
